src/main/java/io/stargate/sgv2/jsonapi/service/operation/model/impl/InsertOperation.java

```diff
@@ -13,61 +14,148 @@
+
+        // concatenate to respect ordered
+        .onItem()
+        .transformToUniAndConcatenate(
+            doc ->
+                insertDocument(queryExecutor, query, doc)
+
+                    // if we fail to insert propagate the FailFastInsertException
+                    .onFailure()
+                    .transform(t -> new FailFastInsertException(doc, t)))
+
+        // if no failures reduce to the op page
+        .collect()
+        .in(InsertOperationPage::new, (agg, in) -> agg.aggregate(in, null))
+
+        // use object identity to resolve to Supplier<CommandResult>
+        .map(i -> i)
+
+        // in case upstream propagated FailFastInsertException
+        // ensure to construct correctly the information about inserted documents
+        .onFailure(FailFastInsertException.class)
+        .recoverWithItem(
+            e -> {
+              // safe to cast, asserted class in onFailure
+              FailFastInsertException failFastInsertException = (FailFastInsertException) e;
+              WritableShreddedDocument failed = failFastInsertException.document;
+
+              // collect inserted, since it's sequential iterate until failed index
+              int failedIndex = documents().lastIndexOf(failed);
+              List<DocumentId> insertedId =
+                  documents().stream()
+                      .limit(failedIndex)
+                      .map(WritableShreddedDocument::id)
+                      .toList();
+
+              return new InsertOperationPage(
+                  insertedId,
+                  Collections.singletonMap(failed.id(), failFastInsertException.getCause()));
+            })
+
+        // use object identity to resolve to Supplier<CommandResult>
+        .map(i -> i);
+  }
+
+  // implementation for the unordered insert
+  private Uni<Supplier<CommandResult>> insertUnordered(QueryExecutor queryExecutor) {
+    // build query once
     QueryOuterClass.Query query = buildInsertQuery();
-    final Uni<List<DocumentId>> ids =
-        Multi.createFrom()
-            .items(documents.stream())
-            .onItem()
-            .transformToUniAndConcatenate(doc -> insertDocument(queryExecutor, query, doc))
-            .collect()
-            .asList();
-    return ids.onItem().transform(insertedIds -> new InsertOperationPage(insertedIds, documents));
+
+    return Multi.createFrom()
+        .iterable(documents)
+
+        // merge to make it parallel
+        .onItem()
+        .transformToUniAndMerge(
+            doc ->
+                insertDocument(queryExecutor, query, doc)
+
+                    // handle errors fail silent mode
+                    .onItemOrFailure()
+                    .transform((id, t) -> Tuple2.of(doc, t)))
+
+        // then reduce here
+        .collect()
+        .in(InsertOperationPage::new, (agg, in) -> agg.aggregate(in.getItem1().id(), in.getItem2()))
+
+        // use object identity to resolve to Supplier<CommandResult>
+        .map(i -> i);
```

Comment:
Why is this needed in() would have already return Uni< InsertOperationPage>

Reply:
Does not compile. The method signature must be 

`Uni<? extends Supplier<CommandResult>>` 

so that we can return the `Uni<InsertOperationPage>`.

I wanted to refactor, but it's not a small change, should be changed in all sub-classes and tests.. Do you think it's worth it?